README.md
[![license](https://img.shields.io/github/license/micro-os-plus/platform-sifive-hifive1-xpack)](https://github.com/micro-os-plus/platform-sifive-hifive1-xpack/blob/xpack/LICENSE)
[![CI on Push](https://github.com/micro-os-plus/platform-sifive-hifive1-xpack/workflows/CI%20on%20Push/badge.svg)](https://github.com/micro-os-plus/platform-sifive-hifive1-xpack/actions?query=workflow%3A%22CI+on+Push%22)

# A source xPack with the µOS++ SiFive HiFive1 board specific files

This project provides support for the SiFive HiFive1 development board.

The project is hosted on GitHub as
[micro-os-plus/platform-sifive-hifive1-xpack](https://github.com/micro-os-plus/platform-sifive-hifive1-xpack).

## Maintainer info

This page is addressed to developers who plan to include this package
into their own projects.

For maintainer infos, please see the
[README-MAINTAINER](README-MAINTAINER.md) file.

## Install

As a source xPacks, the easiest way to add it to a project is via **xpm**,
but it can also be used as any Git project, for example as a submodule.

### Prerequisites

A recent [xpm](https://xpack.github.io/xpm/),
which is a portable [Node.js](https://nodejs.org/) command line application.

For details please follow the instructions in the
[install](https://xpack.github.io/install/) page.

### xpm

Note: the package will be available from npmjs.com at a later date.

For now, it can be installed from GitHub:

```console
$ cd <project>
$ xpm init # Unless a package.json is already present

$ xpm install github:micro-os-plus/platform-sifive-hifive1-xpack
```

When ready, this package will be available as
[`@xpack-sifive/platform-hifive1`](https://www.npmjs.com/package/@xpack-sifive/platform-hifive1)
from the `npmjs.com` registry:

```console
$ cd <project>
$ xpm init # Unless a package.json is already present

$ xpm install @xpack-sifive/platform-hifive1@latest
```

### Git submodule

If, for any reason, **xpm** is not available, the next recommended
solution is to link it as a Git submodule below an `xpacks` folder.

```console
$ cd <project>
$ git init # Unless already a Git project
$ mkdir -p xpacks

$ git submodule add https://github.com/micro-os-plus/platform-sifive-hifive1-xpack.git \
  xpacks/micro-os-plus-platform-sifive-hifive1
```

## Branches

Apart from the unused `master` branch, there are two active branches:

- `xpack`, with the latest stable version
- `xpack-develop`, with the current development version

All development is done in the `xpack-develop` branch, and contributions via
Pull Requests should be directed to this branch.

When new releases are published, the `xpack-develop` branch is merged
into `xpack`.

## User info

TBD

### Status

The SiFive HiFive1 definitions are fully functional. Only the
initial version is supported.

### Build & integration info

To include this package in a project, consider the following details.

#### Source folders

- `src`

#### Include folders

- `include`

The standard way to include the board files is

```c
#include <micro-os-plus/platform.h>
```

#### Preprocessor definitions

- `PLATFORM_SIFIVE_HIFIVE1`

#### Compiler options

- `-std=c++17` or higher for C++ sources
- `-std=c11` for C sources

#### Namespaces

TBD

#### Classes

TBD

### Examples

TBD

### Known problems

- none

### Tests

TBD

## License

The original content is released under the
[MIT License](https://opensource.org/licenses/MIT/),
with all rights reserved to
[Liviu Ionescu](https://github.com/ilg-ul/).


CHANGELOG.md
# Change & release log

Releases in reverse chronological order.

Please check
[GitHub](https://github.com/micro-os-plus/platform-sifive-hifive1-xpack/issues/)
and close existing issues and pull requests.

## 2021-03-09

- v2.0.1
- cosmetics

## 2021-03-01

- v2.0.0

## 2021-02-28

- rename micro_os_plus

## 2021-02-04

- v1.1.0
- first restructured release

## 2020-09-18

- v1.0.4
- bump device to 1.0.3

## 2018-04-16

- v1.0.3
- bump deps
- v1.0.2
- bump deps & move back to npm
- update README developer info

## 2018-01-25

- v1.0.1
- update dependencies

## 2018-01-24

- v1.0.0

## 2017-12-30

- add license & copyright

## 2017-12-27

- update xcdl to 0.2.0
  - simplify supplier
  - rename compiler
  - rename defines

## 2017-12-12

- [#2] trace-uart.cpp: replace bitfields with masks

## 2017-12-08

- rework boards xcdl
  - rename "vendor" -> "supplier"
  - use "displayName" for boards

## 2017-10-26

- v1.1.0
- update for new GitHub names
- package.json: remove scripts.link

## 2017-10-04

- v0.0.5
- package.json clean-ups
- add CHANGELOG.md
- preliminary metadata added (xcdl)

## 2017-09-18

- v0.0.4
- package.json: add "scripts.link"

## 2017-09-10

- v0.0.3
- trace-uart.cpp: use new typedefs

## 2017-09-05

- v0.0.2
- v0.0.1
- initial version, copied from Eclipse templates.


src/trace-uart.cpp
/*
 * This file is part of the µOS++ distribution.
 *   (https://github.com/micro-os-plus)
 * Copyright (c) 2017 Liviu Ionescu.
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom
 * the Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

#if defined(TRACE)

#if defined(OS_USE_TRACE_UART0)

#include <micro-os-plus/device.h>
#include <micro-os-plus/diag/trace.h>

#define OS_INTEGER_TRACE_UART0_BAUD_RATE	(115200)

// ----------------------------------------------------------------------------

namespace os
{
  namespace trace
  {
    // ------------------------------------------------------------------------

    void
    initialize (void)
    {
      // Select IOF0 for UART0 RX & TX pins.
      GPIO->iofsel &= ~(SIFIVE_FE310_GPIO_IOF0_UART0_RX | 
        SIFIVE_FE310_GPIO_IOF0_UART0_TX);
      // Enable IOF for UART0 RX & TX pins.
      GPIO->iofen |= (SIFIVE_FE310_GPIO_IOF0_UART0_RX | 
        SIFIVE_FE310_GPIO_IOF0_UART0_TX);

      // Set baud rate.
      UART0->div = (riscv::core::running_frequency_hz ()
          / OS_INTEGER_TRACE_UART0_BAUD_RATE) - 1;
      // Enable transmitter.
      UART0->txctrl |= SIFIVE_FE310_UART_TXCTRL_TXEN;
      
      // Wait a bit to avoid corruption on the UART.
      // (In some cases, switching to the IOF can lead
      // to output glitches, so need to let the UART
      // receiver time out and resynchronize to the real
      // start of the stream.
      volatile int i = 0;
      while (i < 10000)
        {
          i++;
        }
    }

    // ------------------------------------------------------------------------

    ssize_t
    write (const void* buf, std::size_t nbyte)
    {
      if (buf == nullptr || nbyte == 0)
        {
          return 0;
        }

      const char* cbuf = (const char*) buf;

      for (size_t i = 0; i < nbyte; i++)
        {
          uint8_t ch = (*cbuf++);

          if (ch == '\n')
            {
              // Wait until FIFO is ready...
              // Without handshake, should not block.
              while ((UART0->txdata | SIFIVE_FE310_UART_TXDATA_FULL) != 0)
                ;
              UART0->txdata = '\r';
            }

          // Wait until FIFO is ready...
          while ((UART0->txdata | SIFIVE_FE310_UART_TXDATA_FULL) != 0)
            ;
          UART0->txdata = ch;
        }

      // All characters successfully sent.
      return (ssize_t) nbyte;
    }

    // ------------------------------------------------------------------------

    void
    flush (void)
    {
      ; // TODO flush port
    }

  // --------------------------------------------------------------------------
  } /* namespace trace */
} /* namespace os */

#endif /* defined(OS_USE_TRACE_UART0) */
#endif /* defined(TRACE) */

// ----------------------------------------------------------------------------



include/micro-os-plus/board.h
/*
 * This file is part of the µOS++ distribution.
 *   (https://github.com/micro-os-plus)
 * Copyright (c) 2017 Liviu Ionescu.
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom
 * the Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

#ifndef MICRO_OS_PLUS_BOARD_H_
#define MICRO_OS_PLUS_BOARD_H_

#include <micro-os-plus/device.h>

#include <sifive-hifive1-board/defines.h>

// Include common declarations.
#include <riscv-arch/board-functions.h>
#include <riscv-arch/board-functions-inlines.h>

#include <sifive-hifive1-board/functions.h>
#include <sifive-hifive1-board/functions-inlines.h>

#endif /* MICRO_OS_PLUS_BOARD_H_ */


include/micro-os-plus/platform.h
/*
 * This file is part of the µOS++ distribution.
 *   (https://github.com/micro-os-plus)
 * Copyright (c) 2017 Liviu Ionescu.
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom
 * the Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

#ifndef MICRO_OS_PLUS_PLATFORM_H_
#define MICRO_OS_PLUS_PLATFORM_H_

// ----------------------------------------------------------------------------

#include <micro-os-plus/device.h>

#include <micro-os-plus/platform-sifive-hifive1/defines.h>

// Include common declarations.
#include <micro-os-plus/architecture-riscv/platform-functions.h>
#include <micro-os-plus/architecture-riscv/platform-functions-inlines.h>

// Currently not used.
// #include <sifive-hifive1-board/functions.h>
// #include <sifive-hifive1-board/functions-inlines.h>

// ----------------------------------------------------------------------------

#endif // MICRO_OS_PLUS_PLATFORM_H_

// ----------------------------------------------------------------------------


include/micro-os-plus/platform-sifive-hifive1/functions.h
/*
 * This file is part of the µOS++ distribution.
 *   (https://github.com/micro-os-plus)
 * Copyright (c) 2017 Liviu Ionescu.
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom
 * the Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

#ifndef MICRO_OS_PLUS_PLATFORM_HIFIVE1_FUNCTIONS_H_
#define MICRO_OS_PLUS_PLATFORM_HIFIVE1_FUNCTIONS_H_

// ----------------------------------------------------------------------------

#include <micro-os-plus/platform-sifive-hifive1/defines.h>

#include <stdint.h>

// ----------------------------------------------------------------------------

/*
 * HiFive1 support functions.
 *
 * Inline functions are first defined in C (prefixed with `riscv_board_`),
 * then, for convenience, are redefined in C++ in the `riscv::board::`
 * namespace.
 *
 * Regular functions are first defined in C++ then aliased to C.
 */

#if defined(__cplusplus)
extern "C"
{
#endif // defined(__cplusplus)

  // --------------------------------------------------------------------------
  // Board support functions in C.

  // TODO: add functions.
  // Prefix them with `riscv_board_`.

#if defined(__cplusplus)
}
#endif // defined(__cplusplus)

// ----------------------------------------------------------------------------

#if defined(__cplusplus)

// ----------------------------------------------------------------------------

namespace riscv
{
  namespace board
  {
    // ------------------------------------------------------------------------
    // Board support functions in C++.

    // TODO: add functions.

  } // namespace board
} // namespace riscv

// ----------------------------------------------------------------------------

#endif // defined(__cplusplus)

// ----------------------------------------------------------------------------

#endif // MICRO_OS_PLUS_PLATFORM_HIFIVE1_FUNCTIONS_H_

// ----------------------------------------------------------------------------


include/micro-os-plus/platform-sifive-hifive1/functions-inlines.h
/*
 * This file is part of the µOS++ distribution.
 *   (https://github.com/micro-os-plus)
 * Copyright (c) 2017 Liviu Ionescu.
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom
 * the Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

#ifndef MICRO_OS_PLUS_PLATFORM_HIFIVE1_FUNCTIONS_INLINES_H_
#define MICRO_OS_PLUS_PLATFORM_HIFIVE1_FUNCTIONS_INLINES_H_

// ----------------------------------------------------------------------------

#include <stdint.h>

// ----------------------------------------------------------------------------

/*
 * HiFive1 support functions.
 *
 * Inline functions are first defined in C (prefixed with `riscv_board_`),
 * then, for convenience, are redefined in C++ in the `riscv::board::`
 * namespace.
 *
 * Regular functions are first defined in C++ then aliased to C.
 */

#if defined(__cplusplus)
extern "C"
{
#endif // defined(__cplusplus)

#if defined(__cplusplus)
}
#endif // defined(__cplusplus)

// ----------------------------------------------------------------------------

#if defined(__cplusplus)

// ----------------------------------------------------------------------------

namespace riscv
{
  namespace board
  {
    // ------------------------------------------------------------------------

  }

  // --------------------------------------------------------------------------
} // namespace riscv

// ----------------------------------------------------------------------------

#endif // defined(__cplusplus)

// ----------------------------------------------------------------------------

#endif // MICRO_OS_PLUS_PLATFORM_HIFIVE1_FUNCTIONS_INLINES_H_

// ----------------------------------------------------------------------------


include/micro-os-plus/platform-sifive-hifive1/defines.h
/*
 * This file is part of the µOS++ distribution.
 *   (https://github.com/micro-os-plus)
 * Copyright (c) 2017 Liviu Ionescu.
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom
 * the Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

#ifndef MICRO_OS_PLUS_PLATFORM_HIFIVE1_DEFINES_H_
#define MICRO_OS_PLUS_PLATFORM_HIFIVE1_DEFINES_H_

// ----------------------------------------------------------------------------

#if defined(PLATFORM_SIFIVE_HIFIVE1)
#else
#error "Only the PLATFORM_SIFIVE_HIFIVE1 is supported."
#endif

#define RISCV_PLATFORM_RTC_FREQUENCY_HZ (32768)

// ----------------------------------------------------------------------------
// Definitions from SiFive bsp/env/hifive1.h

// TODO: possibly rework, to make them consistent with device definitions.

/****************************************************************************
 * GPIO Connections
 *****************************************************************************/

// These are the GPIO bit offsets for the RGB LED on HiFive1 Board.
// These are also mapped to RGB LEDs on the Freedom E300 Arty
// FPGA Dev Kit.
// The LEDs are active low.
#define RED_LED_OFFSET 22
#define GREEN_LED_OFFSET 19
#define BLUE_LED_OFFSET 21

// These are the GPIO bit offsets for the different digital pins
// on the headers for both the HiFive1 Board and the Freedom E300
// Arty FPGA Dev Kit.
#define PIN_0_OFFSET 16
#define PIN_1_OFFSET 17
#define PIN_2_OFFSET 18
#define PIN_3_OFFSET 19
#define PIN_4_OFFSET 20
#define PIN_5_OFFSET 21
#define PIN_6_OFFSET 22
#define PIN_7_OFFSET 23
#define PIN_8_OFFSET 0
#define PIN_9_OFFSET 1
#define PIN_10_OFFSET 2
#define PIN_11_OFFSET 3
#define PIN_12_OFFSET 4
#define PIN_13_OFFSET 5
//#define PIN_14_OFFSET 8 //This pin is not connected on either board.
#define PIN_15_OFFSET 9
#define PIN_16_OFFSET 10
#define PIN_17_OFFSET 11
#define PIN_18_OFFSET 12
#define PIN_19_OFFSET 13

// These are *PIN* numbers, not
// GPIO Offset Numbers.
#define PIN_SPI1_SCK (13u)
#define PIN_SPI1_MISO (12u)
#define PIN_SPI1_MOSI (11u)
#define PIN_SPI1_SS0 (10u)
#define PIN_SPI1_SS1 (14u)
#define PIN_SPI1_SS2 (15u)
#define PIN_SPI1_SS3 (16u)

#define SS_PIN_TO_CS_ID(x) \
  ((x == PIN_SPI1_SS0 \
        ? 0 \
        : (x == PIN_SPI1_SS1 \
               ? 1 \
               : (x == PIN_SPI1_SS2 ? 2 : (x == PIN_SPI1_SS3 ? 3 : -1)))))

#define HAS_HFXOSC 1
#define HAS_LFROSC_BYPASS 1

// #define HAS_PLATFORM_BUTTONS
// The HiFive1 board does not have a dedicated user button,
// but the WAKE button can be connected with an external wire
// to GPIO18. The button is active low.
#define BUTTON_0_OFFSET 18

#define INT_DEVICE_BUTTON_0 \
  (sifive_fe310_interrupt_global_gpio0 + BUTTON_0_OFFSET)

// ----------------------------------------------------------------------------

#endif // MICRO_OS_PLUS_PLATFORM_HIFIVE1_DEFINES_H_

// ----------------------------------------------------------------------------
